openwisp_notifications/handlers.py

```diff
@@ -50,8 +54,22 @@ def notify_handler(**kwargs):
     if target_org:
         where = where | (Q(is_staff=True) & Q(openwisp_users_organization=target_org))
         where_group = Q(openwisp_users_organization=target_org)
-    where_group = where_group & Q(notificationuser__receive=True)
-    where = where & Q(notificationuser__receive=True)
+
+    # We can only find notification setting if notification type is present.
+    if notification_type:
+        # Create notification for users who have opted for receiving notifications.
+        # For users who have not configured web_notifications,
+        # use default from notification type
+        web_notification = Q(notificationsetting__web=True)
+        if notification_template['web_notification']:
+            web_notification |= Q(notificationsetting__web=None)
+
+        notification_setting = web_notification & Q(
+            notificationsetting__type=notification_type,
+            notificationsetting__organization_id=target_org,
```

Comment:
since we're querying for these fields, I see that we don't have an index on the type field, can you please investigate what's the best way to add a dedicated index to speed up this query? Either a single index or a composite index, but the query should be fast and be able to scale also when the amount of users in the system is high.

Reply:
I created 100 superusers and 100 organizations with 10 notification types. Following are graphs for 20 cold start(django caches queryset results). Please have a look at this [spreadsheet containing all the data](https://docs.google.com/spreadsheets/d/1N-Z3mQldWJvi7yBgMX7vce2CBTMpNh8N0CA3mJA1q00/edit?usp=sharing) for more information. I think 'type' and 'organization' could be a good candidate, because there will certainly not be as many number of organizations as users in production systems. 


![Look up for users (expected when creating notification)](https://user-images.githubusercontent.com/32094433/90409703-201b5f80-e0c7-11ea-9c1b-3527dcf2d5de.png)
![Lookup for notification settings(expected in admin site) (1)](https://user-images.githubusercontent.com/32094433/90408945-2a892980-e0c6-11ea-8929-f23a0fdb37e2.png)